## Beta-diversity - Technical replicate variation
A validation notebook for comparing technical replicates via performing beta-diversity analysis. Code and example input data are provided for running the analysis for four distance metrics (Jaccard, RPCA, unweighted UniFrac, and weighted UniFrac).

### Dependencies

The first part of this analysis uses a jupyter notebook that uses a QIIME2 conda environment kernel. The second part of this analyses uses R.

### References

This code was adapted from Stephen Wandro (swandro at eng dot ucsd dot edu).

Shaffer, J.P., C.S. Carpenter, C. Martino, R.A. Salido, J.J. Minich, M. Bryant, K. Sanders, T. Schwartz, G. Humphrey, A.D. Swafford, and R. Knight. 2022. A comparison of six DNA extraction protocols for 16S, ITS, and shotgun metagenomic sequencing of microbial communities. Biotechniques 73:34-46. doi: 10.2144/btn-2022-0032.

Shaffer, J.P., C. Marotz, P. Belda-Ferre, C. Martino, S. Wandro, M. Estaki, R.A. Salido, C.S. Carpenter, L.S. Zaramela, J.J. Minich, M. Bryant, K. Sanders, S. Fraraccio, G. Ackmann, G. Humphrey, A. Swafford, S. Miller-Montgomery, and R. Knight. 2020. A comparison of DNA/RNA extraction protocols for high-throughput amplicon and shotgun metagenomics sequencing of microbial communities across diverse sample types. BioTechniques 70:149-159. doi: 10.2144/btn-2020-0153. PMCID: PMC7931620.
